src/id_allocator.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+//! Provides allocation and releasing strategy for IDs.
+//!
+//! This module implements an allocation strategies for all resource types
+//! that can be abstracted to an integer.
+
+use std::collections::BTreeSet;
+use std::fmt;
+use std::result;
+
+// Error type for IdAllocator usage.
+#[derive(Debug)]
+pub enum Error {
+    // All ids from the range specified are exhausted.
+    Overflow,
+    // An id that is not part of the specified range was requested to be released.
+    InvalidInput(u32),
+    // There are no more IDs available in the manage range
+    ResourceExhausted,
+}
+
+pub type Result = result::Result<u32, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            Overflow => write!(f, "Id counter overflowed."),
+            InvalidInput(id) => write!(f, "Specified id: {} is not in the range.", id),
+            ResourceExhausted => {
+                write!(f, "There are no more available ids in the specified range.")
+            }
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # use vm_allocator::IdAllocator;
+///
+/// let mut legacy_irq_allocator = IdAllocator::new(5, 23);
+///
+/// for _ in 1..10 {
+///     let _id = legacy_irq_allocator.allocate_id().unwrap();
+/// }
+/// let irq = 8;
+/// let freed_id_result = legacy_irq_allocator.free_id(irq);
+/// if freed_id_result.is_err() {
+///     println!("Failed to free the specified id: {:?}", freed_id_result);
+/// }
+/// ```
+
+/// Id allocator representation.
+pub struct IdAllocator {
```

Comment:
Can you add a few comments about the internal design choices of this structure? What I am finding difficult to understand is why do we need to use an interval tree as oppose to just a vector in which we keep the freed ids. I would assume that some trade-offs were made and that these advantages and disadvantages were put in balance, but this kind of information tends to be forgotten, so it's better to keep it next to the code. What I don't particularly like about BTree is Rust is that it's making you implement useful operations that are not yet stabilized (the one that would be useful for us is `pop_first`).

Also, the allocation strategy should be mentioned somewhere IMHO (i.e. we first try to reserve IDs from previously freed ones), but I am not sure what is the best way to deliver the message.